Documentation/wip/FunctionVersioningProposal.md

```diff
@@ -0,0 +1,128 @@
+# Function Versioning
+
+## Use cases for function versioning from POV of fission users :
+1. Create different versions of a function with code changes.
+2. List all versions created for a function.
+3. Delete a specific version of a function. (Note : Not allowing entire CRUD on a version of a function. Does update of a function version make any sense? )
+
+4. Create aliases to different function versions. Ex : `"Dev"` referring to `"V3"` of a function, `"Qa"` referring to `"V2"` of a function, `"Prod"` referring to `"V1"` of a function. By default `"latest"` referring to most recent function version created.
+5. List all aliases created for a function.
+6. Update an alias to point to a different version of a function.
+7. Delete a function alias. (This shouldn't delete the function object that this alias refers to, instead just delete the alias object)
+
+8. Delete a function by name, which will then cascade deletion of all aliases and all versions of that particular function.
+
+9. Create an HTTP Trigger for a function with a specific version.
+10. Have two versions of a function deployed and perform canary testing by creating an HTTP trigger for both versions of a function. // TODO : Come back to this
+11. Create other kinds of triggers to a function with a specific version.
+12. Create a group of functions and tag them as one version. ( perhaps run this group of functions as cron jobs with one timeTrigger, haven’t thought this through ) // TODO : Come back to this
+
+
+
+### Design approach :
+#### Data Structures
+Few new data structures for this approach are listed below. The need for these become evident once you go through the GeneralFlow section
+
+* Introducing a new CRD of type `"Alias"`. This will refer to a specific version of a function.
+```
+Type Alias struct {
+	Name string
+	FunctionRef FunctionRef
+} 
+```
+* Introducing a new CRD of type `"FunctionMaster"`. This will have a reference to all aliases of a function. Also will have a list of references to all versions created for this function.
+```
+Type FunctionMaster struct {
+	Name string
+	Alias []Alias
+	Versions []FunctionRef
+}
+```
+
+* Introducing a new CRD of type `"Tag"`. This is to support use case 12 listed above.
+```
+Type Tag struct {
+	Name string
+	FunctionRef []FunctionRef
+}
+```
+
+#### General Flow 
+1. When a user wants to create a function with different versions, he can do so with the following commands
+    ```bash
+    $ fission function version create --name hello --env nodejs --code hello1.js --version v1.0.0
```

Comment:
This is my personal opinion but using a SemVar like versioning (i.e. x.y.z) makes it a bit user un-friendly and might be an overkill. I like simpler naming scheme with a single number. Also in practice how many versions of a function might user create? Even if it is in range of 1000, the single number versioning is far simpler. Open to more discussion on this.

Reply:
Yes, I agree too. But it's upto the user to create versions according to any syntax. I've also listed this point under "Note" section in the end. I'm not very keen on fission maintaining the version values.